overlord/ifacestate/ifacemgr.go

```diff
@@ -261,39 +272,74 @@ func (m *InterfaceManager) autoConnect(task *state.Task, snapName string) error
 	return nil
 }
 
-func (m *InterfaceManager) doRemoveSnapSecurity(task *state.Task, _ *tomb.Tomb) error {
+func (m *InterfaceManager) doRemoveProfiles(task *state.Task, _ *tomb.Tomb) error {
 	task.State().Lock()
 	defer task.State().Unlock()
 
-	// Get snap.Info from bits handed by the snap manager.
-	ss, err := snapstate.TaskSnapSetup(task)
+	snapSetup, err := snapstate.TaskSnapSetup(task)
 	if err != nil {
 		return err
 	}
-	snapInfo, err := snapstate.Info(task.State(), ss.Name, ss.Revision)
+	snapName := snapSetup.Name
+
+	affectedSnaps, err := m.repo.DisconnectSnap(snapName)
 	if err != nil {
 		return err
 	}
-	snapName := snapInfo.Name()
-	affectedSnaps, err := m.repo.DisconnectSnap(snapName)
+	for _, snapInfo := range affectedSnaps {
+		if snapInfo.Name() == snapName {
+			continue
+		}
+		if err := setupSnapSecurity(task, snapInfo, m.repo); err != nil {
+			return state.Retry
+		}
+	}
+
+	if err := m.repo.RemoveSnap(snapName); err != nil {
+		return err
+	}
+	if err := removeSnapSecurity(task, snapName); err != nil {
+		return state.Retry
+	}
+
+	return nil
+}
+
+func (m *InterfaceManager) doDiscardConns(task *state.Task, _ *tomb.Tomb) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	snapSetup, err := snapstate.TaskSnapSetup(task)
 	if err != nil {
 		return err
 	}
-	// TODO: remove all connections from the state
-	if err := m.repo.RemoveSnap(snapName); err != nil {
 
+	snapName := snapSetup.Name
+
+	var snapState snapstate.SnapState
+	err = snapstate.Get(st, snapName, &snapState)
+	if err != nil && err != state.ErrNoState {
 		return err
 	}
-	if len(affectedSnaps) == 0 {
-		affectedSnaps = append(affectedSnaps, snapInfo)
+
+	if err == nil && len(snapState.Sequence) != 0 {
+		return fmt.Errorf("cannot discard connections of a snap %q while it is present", snapName)
 	}
-	for _, snapInfo := range affectedSnaps {
-		for _, backend := range securityBackends {
-			if err := backend.Remove(snapInfo.Name()); err != nil {
-				return state.Retry
-			}
+	conns, err := getConns(st)
+	if err != nil {
+		return err
+	}
```

Comment:
I've implemented a simple undo handler. I'm not sure if that's the right way so have look please.